File: backend/app/routes/projects.py
from uuid import UUID
from typing import List, Optional
from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy import select, func
from sqlalchemy.ext.asyncio import AsyncSession

from app.db.session import get_db
from app.auth.deps import get_current_user
from app.models.project import Project
from app.models.lesson import Lesson
from app.models.user import User
from app.schemas.project import ProjectCreate, ProjectUpdate, ProjectOut
from app.schemas.lesson import LessonOut

router = APIRouter(prefix="/api/projects", tags=["projects"])


@router.get("/", response_model=List[ProjectOut])
async def list_projects(
    search: Optional[str] = Query(None),
    sector: Optional[str] = Query(None),
    status: Optional[str] = Query(None),
    client: Optional[str] = Query(None),
    db: AsyncSession = Depends(get_db),
    _: User = Depends(get_current_user),
):
    q = select(Project)
    if search:
        pattern = f"%{search}%"
        q = q.where(
            Project.project_name.ilike(pattern)
            | Project.client.ilike(pattern)
            | Project.project_number.ilike(pattern)
        )
    if sector:
        q = q.where(Project.sector == sector)
    if status:
        q = q.where(Project.status == status)
    if client:
        q = q.where(func.lower(Project.client) == client.lower())
    q = q.order_by(Project.updated_at.desc())
    result = await db.execute(q)
    return result.scalars().all()


@router.post("/", response_model=ProjectOut, status_code=201)
async def create_project(
    body: ProjectCreate,
    db: AsyncSession = Depends(get_db),
    user: User = Depends(get_current_user),
):
    p = Project(created_by=user.id, updated_by=user.id, **body.model_dump())
    db.add(p)
    await db.commit()
    await db.refresh(p)
    return p


@router.get("/{project_id}", response_model=ProjectOut)
async def get_project(
    project_id: UUID,
    db: AsyncSession = Depends(get_db),
    _: User = Depends(get_current_user),
):
    result = await db.execute(select(Project).where(Project.id == project_id))
    p = result.scalar_one_or_none()
    if not p:
        raise HTTPException(404, "Project not found")
    return p


@router.patch("/{project_id}", response_model=ProjectOut)
async def update_project(
    project_id: UUID,
    body: ProjectUpdate,
    db: AsyncSession = Depends(get_db),
    user: User = Depends(get_current_user),
):
    result = await db.execute(select(Project).where(Project.id == project_id))
    p = result.scalar_one_or_none()
    if not p:
        raise HTTPException(404, "Project not found")
    for field, value in body.model_dump(exclude_unset=True).items():
        setattr(p, field, value)
    p.updated_by = user.id
    await db.commit()
    await db.refresh(p)
    return p


@router.delete("/{project_id}", status_code=204)
async def delete_project(
    project_id: UUID,
    db: AsyncSession = Depends(get_db),
    _: User = Depends(get_current_user),
):
    result = await db.execute(select(Project).where(Project.id == project_id))
    p = result.scalar_one_or_none()
    if not p:
        raise HTTPException(404, "Project not found")
    await db.delete(p)
    await db.commit()


@router.get("/{project_id}/lessons", response_model=List[LessonOut])
async def get_project_lessons(
    project_id: UUID,
    db: AsyncSession = Depends(get_db),
    _: User = Depends(get_current_user),
):
    result = await db.execute(
        select(Lesson).where(Lesson.project_id == project_id).order_by(Lesson.updated_at.desc())
    )
    return result.scalars().all()

File: frontend/src/api/projects.ts
import api from "./client";

export interface Project {
  id: string;
  project_number: string | null;
  project_name: string;
  client: string | null;
  location: string | null;
  contract_value: number | null;
  year_completed: string | null;
  status: string;
  wsp_role: string | null;
  project_manager: string | null;
  sector: string | null;
  services_performed: string | null;
  key_personnel: { name: string; role: string }[];
  description: string | null;
  outcomes: string | null;
  updated_at: string | null;
}

export const projectsApi = {
  list: (params?: { search?: string; sector?: string; status?: string; client?: string }) =>
    api.get<Project[]>("/api/projects/", { params }).then(r => r.data),
  get: (id: string) =>
    api.get<Project>(`/api/projects/${id}`).then(r => r.data),
  create: (data: Partial<Project>) =>
    api.post<Project>("/api/projects/", data).then(r => r.data),
  update: (id: string, data: Partial<Project>) =>
    api.patch<Project>(`/api/projects/${id}`, data).then(r => r.data),
  delete: (id: string) =>
    api.delete(`/api/projects/${id}`),
  lessons: (id: string) =>
    api.get<any[]>(`/api/projects/${id}/lessons`).then(r => r.data),
};
